models/imap.go

```diff
@@ -0,0 +1,152 @@
+var ErrIMAPPortNotSpecified = errors.New("No IMAP Port specified")
+
+// ErrInvalidIMAPHost indicates that the IMAP server string is invalid
+var ErrInvalidIMAPHost = errors.New("Invalid IMAP server address")
+
+// ErrInvalidIMAPPort indicates that the IMAP Port is invalid
+var ErrInvalidIMAPPort = errors.New("Invalid IMAP Port")
+
+// ErrIMAPUsernameNotSpecified is thrown when there is no Username specified
+// in the IMAP configuration
+var ErrIMAPUsernameNotSpecified = errors.New("No Username specified")
+
+// ErrIMAPPasswordNotSpecified is thrown when there is no Password specified
+// in the IMAP configuration
+var ErrIMAPPasswordNotSpecified = errors.New("No Password specified")
+
+var ErrInvalidIMAPFreq = errors.New("Invalid polling frequency.")
+
+// TableName specifies the database tablename for Gorm to use
+func (s IMAP) TableName() string {
+	return "imap"
+}
+
+// Validate ensures that IMAP configs/connections are valid
+func (s *IMAP) Validate() error {
+	switch {
+	case s.Host == "":
+		return ErrIMAPHostNotSpecified
+	case s.Port == 0:
+		return ErrIMAPPortNotSpecified
+	case s.Username == "":
+		return ErrIMAPUsernameNotSpecified
+	case s.Password == "":
+		return ErrIMAPPasswordNotSpecified
+	}
+
+	// Set the default value for Folder
+	if s.Folder == "" {
+		s.Folder = DefaultIMAPFolder
+	}
+
+	// Make sure s.Host is an IP or hostname. NB will fail if unable to resolve the hostname.s
+	ip := net.ParseIP(s.Host)
+	_, err := net.LookupHost(s.Host)
+	if ip == nil && err != nil {
+		return ErrInvalidIMAPHost
+	}
+
+	// Make sure 1 >= port <= 65535
+	if s.Port < 1 || s.Port > 65535 {
+		return ErrInvalidIMAPPort
+	}
+
+	// Make sure the polling frequency is between every 30 seconds and every year
+	// If not set it to the default
+	if s.IMAPFreq < 30 || s.IMAPFreq > 31540000 {
+		s.IMAPFreq = DefaultIMAPFreq
+	}
+
+	return nil
+}
+
+// GetIMAP returns the IMAP server owned by the given user.
+func GetIMAP(uid int64) ([]IMAP, error) {
+	ss := []IMAP{}
+	count := 0
+	err := db.Where("user_id=?", uid).Find(&ss).Count(&count).Error
+
+	if err != nil {
+		log.Error(err)
+		return ss, err
+	}
+	return ss, nil
+}
+
+// PostIMAP updates IMAP settings for a user in the database.
+func PostIMAP(s *IMAP, uid int64) error {
+	err := s.Validate()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	// Delete old entry. TODO: Save settings and if fails to Save below replace with original
+	err = DeleteIMAP(uid)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	// Insert new settings into the DB
+	err = db.Save(s).Error
+	if err != nil {
+		log.Error("Bad things happened here ", err.Error())
+	}
+	return err
+}
+
+// DeleteIMAP deletes the existing IMAP in the database.
+func DeleteIMAP(uid int64) error {
+	err := db.Where("user_id=?", uid).Delete(&IMAP{}).Error
+	if err != nil {
+		log.Error(err)
+	}
+	return err
+}
+
+func SuccessfulLogin(s *IMAP) error {
+	err := db.Model(&s).Update("last_login", time.Now()).Error
```

Comment:
```suggestion
	err := db.Model(&s).Update("last_login", time.Now().UTC()).Error
```

We try to use UTC everywhere else, so it'd be good to keep things consistent here.

Reply:
Done ✅